allennlp/models/constituency_parser.py

```diff
@@ -249,21 +252,152 @@ def decode(self, output_dict: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor
         all_sentences = output_dict["tokens"]
         all_pos_tags = output_dict["pos_tags"] if all(output_dict["pos_tags"]) else None
         num_spans = output_dict["num_spans"].data
-        trees = self.construct_trees(all_predictions, all_spans, num_spans, all_sentences, all_pos_tags)
 
         batch_size = all_predictions.size(0)
         output_dict["spans"] = [all_spans[i, :num_spans[i]] for i in range(batch_size)]
         output_dict["class_probabilities"] = [all_predictions[i, :num_spans[i], :] for i in range(batch_size)]
 
-        output_dict["trees"] = trees
+        top_k_trees = self.construct_topk_trees(num_trees=16,
+                                                predictions=all_predictions,
+                                                all_spans=all_spans,
+                                                num_spans=num_spans,
+                                                sentences=all_sentences,
+                                                pos_tags=all_pos_tags)
+
+        output_dict["trees"] = [trees[0] for trees in top_k_trees]
+        output_dict['top_k_trees'] = top_k_trees
         return output_dict
 
-    def construct_trees(self,
+    def compute_k_best(self,
+                       sentence: List[str],
+                       pos_tags: List[str],
+                       label_probabilities: torch.FloatTensor,
+                       span_to_index: Dict[Tuple[int, int], int],
+                       num_trees: List[int],
+                       distinguish_between_labels: bool = False):
+        """
+        :param sentence: The sentence for which top-k parses are being computed.
+        :param pos_tags: Part of speech tags for every token in the input sentence.
+        :param label_probabilities: A numpy array of shape (num_spans, num_labels)
+        :param span_to_index: A dictionary mapping span indices to column indices in
+        label_log_probabilities.
+        :param no_label_id: The id of the empty label.
+        :param num_trees: The number of parses required.
+        :param distinguish_between_labels: Whether to distinguish between different labels for the
+        top k trees.
+        :return: A list of the num_tree parses, and their log probabilities.
+        """
+        empty_label_index = self.vocab.get_token_index("NO-LABEL", "labels")
+        all_labels = [self.vocab.get_token_from_index(index, "labels").split("-") for index in
```

Comment:
This can just be `[label.split("-") for label in self.vocab.get_token_to_index_vocabulary("labels").keys()]`

Also, i'd prefer it if you moved the splitting on `-` to inside the tree construction code as it was before. Having it here makes it very unclear why we would want to split the label up. Also, add back my comment for why we did the splitting.